pkg/kubelet/cm/dra/claiminfo.go

```diff
@@ -22,33 +22,16 @@ import (
 
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/kubelet/cm/dra/state"
 	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
 )
 
 // claimInfo holds information required
 // to prepare and unprepare a resource claim.
 type claimInfo struct {
 	sync.RWMutex
-
-	// name of the DRA driver
-	driverName string
-
-	// claimUID is an UID of the resource claim
-	claimUID types.UID
-
-	// claimName is a name of the resource claim
-	claimName string
-
-	// namespace is a claim namespace
-	namespace string
-
-	// podUIDs is a set of pod UIDs that reference a resource
-	podUIDs sets.Set[string]
-
-	// cdiDevices is a list of CDI devices returned by the
-	// GRPC API call NodePrepareResource
-	cdiDevices []string
-
+	state.ClaimInfoState
 	// annotations is a list of container annotations associated with
 	// a prepared resource
 	annotations []kubecontainer.Annotation
```

Comment:
We don't checkpoint annotations, but `Manager.GetResources` relies on them. This can cause empty annotations passed to the CRI runtime if Kubelet restarts between successful `PrepareResources` and `GetResources` calls.
I'd propose to get rid of this field and generate annotations on the fly.

Reply:
but we regenerate them when kubelet restart in the newClaimInfoCache by calling newClaimInfo

Reply:
I don't quite follow. We always generate the annotations anytime a claiminfo is created -- whether that be directly after a `PrepareResources()` call or after a kubelet restart when repopulating the cache.

Reply:
It seems that annotations are generated when Kubelet restarts. Looks quite implicit to me.

Reply:
So, we have 2 parts of cache. One part is checkpointed and another is generated, correct?

Reply:
correct

Reply:
We have a single cache that contains ClaimInfo objects. ClaimInfo objects contain (1) state that is checkpointed and (2) meta information that can be generated from that checkpointed state (for convenience). Right now this meta information is just the annotations but could be other things later on.

From the perspective of using the cache, all anyone needs to know is that it contains ClaimInfo objects and that these objects are always fully populated with the data they store.

Reply:
And we have to remember to re-generate meta information when we restore state, correct?

Reply:
It's not about remembering -- its all encapsulated in the body of the `newClaimInfo()` call. Just as with any `new*()` call you do what's necessary to populate your fields.

Reply:
Now it started to make sense to me. Thanks for the explanations!

Reply:
No worries. Thanks for probing. Keeps us honest.